Define new args structure for registering speed bump

goldenapples opened the pull request:
See #36.

Work in progress - needs more abstraction and fleshing out, as well as a lot of tests to support it. My goal is to build a structure where all of the old arguments will still work as intended by mapping to new arguments. In places where that would introduce too much complexity - and maybe elsewhere, seeing as we're the only ones using this so far, it might be OK to break backward compatibility and just ship theme changes along with the submodule update.

danielbachhuber commented:
> My goal is to build a structure where all of the old arguments will still work as intended by mapping to new arguments. In places where that would introduce too much complexity - and maybe elsewhere, seeing as we're the only ones using this so far, it might be OK to break backward compatibility and just ship theme changes along with the submodule update.

Let's just make breaking changes. We haven't shipped a release of the open source plugin yet, so I'd be fine changing the arguments.

goldenapples commented:
> Let's just make breaking changes.

Sounds good to me.

goldenapples commented:
@noppanit or @danielbachhuber - can you look over this general structure and let me know if you have ideas for improving on it?

This is just a first pass - everything in here needs to be supported by tests, and I fully expect to find a lot of bugs once I start building out the test for it, but I'd like to get a sense of whether this structure seems like it will meet our needs and if it looks extensible and logical enough...

danielbachhuber commented on speed-bumps.php:
```diff
@@ -132,33 +146,100 @@ public function insert_speed_bumps( $the_content ) {
+			// Positional rules: distance from start, end, and other elements
+			'from_start' => array(
+				'paras' => 3,
+				'words' => 75,
```
Is this an `OR` comparison or `AND` ?

goldenapples replied:
I'm treating it as an `AND`. Every one of those conditions has to be met in order to insert a speed bump. This approach should address conditions where the first three paragraphs are short sentences. So a speed bump can't be inserted until at least 75 words in, which could be three paragraphs or 5 or more...

danielbachhuber replied:
Cool. Can we use `paragraphs` instead of `paras`, and `characters` instead of `chars`? The former is more precise.

danielbachhuber commented:
:100: Love it

goldenapples commented:
OK, all tests pass now, and I added coverage for a lot of the new constraints . Fixed a bunch of bugs and inconsistencies in implementation. @noppanit @danielbachhuber - either of you want to review this and give me feedback on it?

noppanit commented:
The code looks good to me. I think this one is related to this GH https://github.com/fusioneng/speed-bumps/issues/5

danielbachhuber commented on speed-bumps.php:
```diff
@@ -132,33 +146,100 @@ public function insert_speed_bumps( $the_content ) {
+				//'image',
+				//),
+			//'minimum_space_from_other_inserts' => 1,
+			//);
```
Are we dropping this?

danielbachhuber commented:
@goldenapples Looks good. Can you:

1. Resolve the merge conflict
2. Merge to master
3. Set up a PR for the theme which cleans up our existing speed bumps?

goldenapples commented:
Merged master with the latest tests and made them all pass. Merging this now, and will follow with an update in fusion theme to use the new args structure.
